File: 프로그래머스/2/87946. 피로도/피로도.py
visited = []
ans = 0
def solution(k, dungeons):
    global visited, ans
    visited = [False] * len(dungeons)
    ans = 0
    
    dfs(k, dungeons, 0)
    return ans

def dfs(k, dungeons, cnt):
    global visited, ans
    ans = max(ans, cnt)
    
    for i in range(len(dungeons)):
        if not visited[i] and k >= dungeons[i][0]:
            visited[i] = True
            dfs(k - dungeons[i][1], dungeons, cnt+1)
            visited[i] = False
    return
        

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static int ans;
    static boolean[] visited;
    public int solution(int k, int[][] dungeons) {
        //k = 현재 피로도
        //dungenons -> 최소 필요 피로도, 소모 피로도
        ans = Integer.MIN_VALUE;
        visited = new boolean[dungeons.length];
        
        dfs(0, k, dungeons);
        return ans;
    }
    static void dfs(int cnt, int k, int[][] dungeons){
        ans = Math.max(ans, cnt);
        
        for(int i=0; i<dungeons.length; i++){
            if(!visited[i]){
                if(k >= dungeons[i][0]){
                    visited[i] = true;
                    dfs(cnt + 1, k - dungeons[i][1], dungeons);
                    visited[i] = false;
                }
            }
        }
    }
}